### PR TITLE
Google reg v0.1

### PR DESCRIPTION
**Pull request by vahushty**

### CODE REVIEW

File: backend/src/services/authGoogleService.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import { OAuth2Client } from "google-auth-library";
+import { prisma } from "@/database/db";
+import jwt from "jsonwebtoken";
+import { v4 } from "uuid";
+
+const CLIENT_ID =
+  "133300448301-7jr985i9aljbqdbf27okphk58b2f987n.apps.googleusercontent.com";
+const CLIENT_SECRET = "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";
+const REDIRECT_URI = "http://localhost:3008/register/google/callback";
```
Comment (SlavikTraktor) [inline backend/src/services/authGoogleService.ts]: It is better to put these constants in .env file

File: backend/src/services/authGoogleService.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  return url;
+};
+
+export const getGoogleTokenAndReg = async (code: unknown) => {
+  const { tokens } = await oAuth2Client.getToken(code as string);
+  oAuth2Client.setCredentials({
+    access_token: tokens.access_token,
+    refresh_token: tokens.refresh_token,
+  });
+  interface GoogleUser {
```
Comment (SlavikTraktor) [inline backend/src/services/authGoogleService.ts]: Lets always create interfaces on top level

File: backend/src/services/authGoogleService.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    email: string;
+    name: string;
+    picture: string;
+  }
+  const { data } = await oAuth2Client.request<GoogleUser>({
```
Comment (SlavikTraktor) [inline backend/src/services/authGoogleService.ts]: Please move this request to separate file and put this file inside `backend/src/requests/googleRequests` folder. (with interface above)

File: backend/src/services/authGoogleService.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import { prisma } from "@/database/db";
+import jwt from "jsonwebtoken";
+import { v4 } from "uuid";
+
+const CLIENT_ID =
+  "133300448301-7jr985i9aljbqdbf27okphk58b2f987n.apps.googleusercontent.com";
+const CLIENT_SECRET = "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";
+const REDIRECT_URI = "http://localhost:3008/register/google/callback";
+
+const oAuth2Client = new OAuth2Client(CLIENT_ID, CLIENT_SECRET, REDIRECT_URI);
```
Comment (SlavikTraktor) [inline backend/src/services/authGoogleService.ts]: it is better to move this `oAuth2Client` to separate file in `backend/src/clients/google` folder.

File: backend/src/services/authGoogleService.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  interface GoogleUser {
+    email: string;
+    name: string;
+    picture: string;
+  }
+  const { data } = await oAuth2Client.request<GoogleUser>({
+    url: "https://www.googleapis.com/oauth2/v3/userinfo",
+  });
+
+  const userF = await prisma.user.findFirst({
```
Comment (SlavikTraktor) [inline backend/src/services/authGoogleService.ts]: Please rename it to just `user`

File: backend/src/services/authGoogleService.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    where: {
+      email: data.email,
+    },
+  });
+  if (!userF) {
+    const userC = await prisma.user.create({
```
Comment (SlavikTraktor) [inline backend/src/services/authGoogleService.ts]: It would be better to rename it to `newUser`

File: backend/src/services/authGoogleService.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        name: data.name,
+        googleReg: true,
+      },
+    });
+    return generateNewTokenPair(userC.id);
+  }
+  return generateNewTokenPair(userF.id);
+};
+
+const generateNewTokenPair = async (userId: number) => {
```
Comment (SlavikTraktor) [inline backend/src/services/authGoogleService.ts]: It is bad practice to copy already created code, just import `generateNewTokenPair` method from authService

File: backend/src/services/authService.ts [hunk 8]
```diff
@@ -13,8 +13,12 @@ export const loginUser = async (name: string, pass: string) => {
   if (!user) {
     throw new Error("User not found");
   }
+  
+  if (user.googleReg == true) {
+    throw new Error("Login through Google");
+  }
 
-  const validatePassword = await bcrypt.compare(pass, user.password);
+  const validatePassword = await bcrypt.compare(pass, user.password as string);
```
Comment (SlavikTraktor) [inline backend/src/services/authService.ts]: it is better to throw error above if password is null instead of `as string`

**Comment by vahushty** [comment]
[[BE] Registration from Google](https://trello.com/c/0OefHae1/9-be-registration-from-google)

File: backend/.env.example [hunk 9]
```diff
@@ -1,6 +1,10 @@
...
\ No newline at end of file
+APP_ENV=dev
+APP_PORT=3008
+AUTH_SECRET=auth_secret_key
+
+
+DATABASE_URL="file:./somename.db"
+
+CLIENT_ID = 133300448301-7jr985i9aljbqdbf27okphk58b2f987n.apps.googleusercontent.com
+CLIENT_SECRET = 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
```
Comment (SlavikTraktor) [inline backend/.env.example]:
```
CLIENT_ID = string
CLIENT_SECRET = string
```

We dont want to share private data to every user on github who has access to this repo)

File: backend/.env.test [hunk 10]
```diff
@@ -1,5 +1,9 @@
...
-
-DATABASE_URL="file:./test.db"
+APP_ENV=test
+APP_PORT=3008
+AUTH_SECRET=some_secret
+
+DATABASE_URL="file:./test.db"
+
+CLIENT_ID = 133300448301-7jr985i9aljbqdbf27okphk58b2f987n.apps.googleusercontent.com
+CLIENT_SECRET = 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
```
Comment (SlavikTraktor) [inline backend/.env.test]: same as above

**Comment by vahushty** [comment]
I fixed everything you asked for

**Comment by vahushty** [comment]
Fixed all again

File: package.json [hunk 11]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "discord-oauth2": "^2.11.0"
```
Comment (SlavikTraktor) [inline package.json]: this file and package-lock in root should be deleted) last fix

**Review by SlavikTraktor — APPROVED** [review]

**Pull request merged**